text/0000-int-guidelines.md

```diff
@@ -0,0 +1,164 @@
+- Start Date: 2015-01-26
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes usage guidelines for the various integer types.
+
+# Motivation
+
+The goal of this RFC is to help people decide what integer types to use when they need
+to make a decision for a new API.
+
+It builds on [https://github.com/rust-lang/rfcs/pull/560](the integer overflow RFC),
+which provides debug-time assertions for overflow and underflow. One of the goals
+of this RFC is to provide guidance for when to use unsigned types, and these
+assertions affect the traditional tradeoffs about unsigned types, which are described
+under the detailed design below.
+
+It also draws inspiration from the [Google C++ Style Guide](http://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Integer_Types)
+
+This RFC attempts to balance several concerns:
+
+* Rust developers should have an easy, go-to heuristic for deciding what
+  integer sizes to use when they need to use a number type.
+* 32-bit integers are considerably faster than 64-bit integers in some
+  situations, so using 64-bit integers for tiny numbers, especially in
+  hot code, can result in unnecessarily slower programs.
+* Reflexive use of 32-bit integers too-often results in overflows when the
+  numbers aren't expected to be "laughably smaller" than the maximum
+  32-bit number.
+* Occasional use of `usize` when building a brand new data structure may
+  be appropriate, but use of `usize` in general can introduce portability
+  hazards when the use-case is not proportional to the amount of
+  addressable memory.
+* Using of unsigned integers is traditionally thought to be error-prone, and
+  style guides often suggest avoiding them. That said, Rust's unsigned integers
+  have built-in underflow assertions, which changes the analysis.
+
+# Detailed design
+
+When deciding what integer type to use for a given situation, you can make the decision
+to a first approximation through this heuristic:
+
+* If you expect all uses of this API to use numbers "laughably smaller" then 32-bits,
+  use 32-bit numbers. Otherwise, prefer 64-bit numbers.
+* If you are sure that the number will never be less than 0, use unsigned integers.
+  Otherwise, use signed numbers.
+* If you are building a new data structure, and your number refers to the size of memory,
```

Comment:
Suggestion: Make this more precise as "... your number is bounded by the number of bytes in memory".